File: packages/lhc/src/domains/inspect/internal/health.ts
// Health composition (Flow 4): both owners' report surfaces joined into
// state counts, actionable failure detail, a repair preview, and live queue
// visibility. Assembled ENTIRELY from FormReportEntry rows — never a
// derived_form or work_item read (the must-not-own rule is this domain's
// whole contract) — and never executing what it previews (AC-4.3).
import type { FormReportEntry } from "../../../shared/derivation.js";
import type { OpResult } from "../../../shared/errors.js";
import type { HealthReport } from "../../../shared/inspect.js";
import * as messages from "../../messages/index.js";
import * as turns from "../../turns/index.js";
import type { ThreadRef } from "../../threads/index.js";

type Owner = "messages" | "turns";

function emptyCounts(): HealthReport["owners"][number]["counts"] {
  return { ready: 0, pending: 0, retrying: 0, failed: 0, blocked: 0 };
}

// Failure detail (AC-4.2): attempts and last error read from wherever the
// mechanics durably put them — retry exhaustion copies them onto the form's
// metadata before the queue row is deleted; a still-live item carries them
// on the queue join. Never synthesized.
function failureOf(
  owner: Owner,
  entry: FormReportEntry,
): HealthReport["failures"][number] {
  const failure: HealthReport["failures"][number] = {
    owner,
    subjectKind: entry.subjectKind,
    subjectId: entry.subjectId,
    form: entry.form,
    reason: entry.reason ?? "",
    attempts: entry.metadata?.attempts ?? entry.queue?.attempts ?? 0,
  };
  const lastError = entry.metadata?.lastError ?? entry.queue?.lastError;
  if (lastError !== undefined) failure.lastError = lastError;
  return failure;
}

export async function composeHealth(ref: ThreadRef): Promise<OpResult<HealthReport>> {
  const messageReport = await messages.report(ref);
  if (!messageReport.ok) return messageReport;
  const turnReport = await turns.report(ref);
  if (!turnReport.ok) return turnReport;

  const sources: ReadonlyArray<readonly [Owner, readonly FormReportEntry[]]> = [
    ["messages", messageReport.value],
    ["turns", turnReport.value],
  ];

  const countsByOwnerKind = new Map<string, HealthReport["owners"][number]>();
  const failures: HealthReport["failures"] = [];
  const repairPreview: HealthReport["repairPreview"] = [];
  const queue = { queued: 0, claimed: 0 };

  for (const [owner, entries] of sources) {
    for (const entry of entries) {
      const key = `${owner}:${entry.form}`;
      let row = countsByOwnerKind.get(key);
      if (row === undefined) {
        row = { owner, kind: entry.form, counts: emptyCounts() };
        countsByOwnerKind.set(key, row);
      }
      switch (entry.state) {
        case "ready":
          row.counts.ready += 1;
          break;
        case "pending":
          if ((entry.queue?.attempts ?? 0) > 0) row.counts.retrying += 1;
          else row.counts.pending += 1;
          break;
        case "failed":
          row.counts.failed += 1;
          failures.push(failureOf(owner, entry));
          // The preview is exactly the failed-and-not-blocked set: blocked is
          // a distinct state, so failed entries ARE the requeue targets.
          repairPreview.push({
            owner,
            subjectKind: entry.subjectKind,
            subjectId: entry.subjectId,
            form: entry.form,
          });
          break;
        case "blocked":
          row.counts.blocked += 1;
          failures.push(failureOf(owner, entry));
          break;
      }
      // Live queue visibility (AC-4.5), per report entry: every pending or
      // retrying entry rides a live item, so queued + claimed here equals
      // pending + retrying above by construction.
      if (entry.queue !== undefined) {
        if (entry.queue.status === "queued") queue.queued += 1;
        else queue.claimed += 1;
      }
    }
  }

  // Deterministic order: messages before turns, kinds alphabetical within
  // an owner — repeated reads with no writes between are deep-equal.
  const owners = [...countsByOwnerKind.values()].sort((a, b) =>
    a.owner === b.owner ? a.kind.localeCompare(b.kind) : a.owner.localeCompare(b.owner),
  );

  return { ok: true, value: { owners, failures, repairPreview, queue } };
}

File: packages/lhc-rs/src/inspect/internal/health.rs
//! Ported from packages/lhc/src/inspect/internal/health.ts.
//!
//! Health composition joins owners' report surfaces into state counts,
//! actionable failure detail, a repair preview, and live queue visibility.
//! Message/turn derivation health comes from DerivationReportEntry rows; capture
//! gaps come from durable source-event markers recorded by capture.

use std::collections::HashMap;

use crate::intake_stream;
use crate::intake_stream::EventRecord;
use crate::messages;
use crate::shared_tech::derivation::{DerivationReportEntry, DerivationState, QueueStatus};
use crate::shared_tech::errors::OpResult;
use crate::shared_tech::inspect::{
    HealthFailure, HealthOwner, HealthOwnerCounts, HealthOwnerEntry, HealthQueue,
    HealthRepairPreview, HealthReport,
};
use crate::threads::ThreadRef;
use crate::turns;

/// TS private `type Owner = "capture" | "messages" | "turns"`.
type Owner = HealthOwner;

/// Capture-gap runtime_note text prefix — byte-exact from TS.
const CAPTURE_GAP_TEXT_PREFIX: &str = "capture gap:";

/// Capture-gap owner/kind key and kind label — byte-exact from TS.
const CAPTURE_GAP_OWNER_KIND_KEY: &str = "capture:capture_gap";
const CAPTURE_GAP_KIND: &str = "capture_gap";

/// Capture-gap failure subjectKind — byte-exact from TS.
const CAPTURE_GAP_SUBJECT_KIND: &str = "event";

fn empty_counts() -> HealthOwnerCounts {
    HealthOwnerCounts {
        ready: 0,
        pending: 0,
        failed: 0,
        blocked: 0,
    }
}

fn capture_gap_text(event: &EventRecord) -> Option<String> {
    let EventRecord::RuntimeNote { payload, .. } = event else {
        return None;
    };
    if payload.text.starts_with(CAPTURE_GAP_TEXT_PREFIX) {
        Some(payload.text.clone())
    } else {
        None
    }
}

fn failure_of(owner: Owner, entry: &DerivationReportEntry) -> HealthFailure {
    HealthFailure {
        owner: owner.as_str().to_string(),
        subject_kind: entry.subject_kind.as_str().to_string(),
        subject_id: entry.subject_id.clone(),
        derivation_type: entry.derivation_type.clone(),
        // TS: entry.reason ?? ""
        reason: entry.reason.clone().unwrap_or_default(),
    }
}

pub async fn compose_health(ref_: ThreadRef) -> OpResult<HealthReport> {
    let message_entries = match messages::report(ref_.clone(), None).await {
        OpResult::Ok { value } => value,
        OpResult::Err { error } => return OpResult::Err { error },
    };
    let turn_entries = match turns::report(ref_.clone(), None).await {
        OpResult::Ok { value } => value,
        OpResult::Err { error } => return OpResult::Err { error },
    };
    let events = match intake_stream::list_events(ref_).await {
        OpResult::Ok { value } => value,
        OpResult::Err { error } => return OpResult::Err { error },
    };

    let sources: [(Owner, &[DerivationReportEntry]); 2] = [
        (HealthOwner::Messages, &message_entries),
        (HealthOwner::Turns, &turn_entries),
    ];

    let mut counts_by_owner_kind: HashMap<String, HealthOwnerEntry> = HashMap::new();
    let mut failures: Vec<HealthFailure> = Vec::new();
    let mut repair_preview: Vec<HealthRepairPreview> = Vec::new();
    let mut queue = HealthQueue {
        queued: 0,
        claimed: 0,
    };

    let capture_gaps: Vec<(&EventRecord, String)> = events
        .iter()
        .filter_map(|event| capture_gap_text(event).map(|text| (event, text)))
        .collect();
    if !capture_gaps.is_empty() {
        let mut counts = empty_counts();
        counts.failed = capture_gaps.len() as i64;
        counts_by_owner_kind.insert(
            CAPTURE_GAP_OWNER_KIND_KEY.to_string(),
            HealthOwnerEntry {
                owner: HealthOwner::Capture,
                kind: CAPTURE_GAP_KIND.to_string(),
                counts,
            },
        );
        for (event, text) in capture_gaps {
            failures.push(HealthFailure {
                owner: HealthOwner::Capture.as_str().to_string(),
                subject_kind: CAPTURE_GAP_SUBJECT_KIND.to_string(),
                subject_id: event.event_order().to_string(),
                derivation_type: CAPTURE_GAP_KIND.to_string(),
                reason: text,
            });
        }
    }

    for (owner, entries) in sources {
        for entry in entries {
            let key = format!("{}:{}", owner.as_str(), entry.derivation_type);
            let row = counts_by_owner_kind
                .entry(key)
                .or_insert_with(|| HealthOwnerEntry {
                    owner,
                    kind: entry.derivation_type.clone(),
                    counts: empty_counts(),
                });
            match entry.state {
                DerivationState::Ready => row.counts.ready += 1,
                DerivationState::Pending => row.counts.pending += 1,
                DerivationState::Failed => {
                    row.counts.failed += 1;
                    failures.push(failure_of(owner, entry));
                    // The preview is exactly the failed-and-not-blocked set: blocked is
                    // a distinct state, so failed entries ARE the requeue targets.
                    repair_preview.push(HealthRepairPreview {
                        owner: owner.as_str().to_string(),
                        subject_kind: entry.subject_kind.as_str().to_string(),
                        subject_id: entry.subject_id.clone(),
                        derivation_type: entry.derivation_type.clone(),
                    });
                }
                DerivationState::Blocked => {
                    row.counts.blocked += 1;
                    failures.push(failure_of(owner, entry));
                }
            }
            // Live queue visibility, per report entry: every pending entry rides a
            // live item, so queued + claimed here equals pending above by
            // construction. Counts are per derivation-report entry:
            // one work item may back multiple entries, so a raw work-item count would
            // break that identity.
            if let Some(q) = &entry.queue {
                match q.status {
                    QueueStatus::Queued => queue.queued += 1,
                    QueueStatus::Claimed => queue.claimed += 1,
                }
            }
        }
    }

    // Deterministic order: messages before turns, kinds alphabetical within
    // an owner — repeated reads with no writes between are deep-equal.
    let mut owners: Vec<HealthOwnerEntry> = counts_by_owner_kind.into_values().collect();
    owners.sort_by(|a, b| {
        a.owner
            .as_str()
            .cmp(b.owner.as_str())
            .then_with(|| a.kind.cmp(&b.kind))
    });

    OpResult::Ok {
        value: HealthReport {
            owners,
            failures,
            repair_preview,
            queue,
        },
    }
}
